src/mlpack/methods/ann/rbm/spike_slab_rbm_policy_impl.hpp

```diff
@@ -0,0 +1,341 @@
+
+#ifndef MLPACK_METHODS_ANN_RBM_SPIKE_SLAB_RBM_POLICY_IMPL_HPP
+#define MLPACK_METHODS_ANN_RBM_SPIKE_SLAB_RBM_POLICY_IMPL_HPP
+
+#include "spike_slab_rbm_policy.hpp"
+
+namespace mlpack {
+namespace ann {
+
+inline SpikeSlabRBMPolicy::SpikeSlabRBMPolicy(const size_t visibleSize,
+      const size_t hiddenSize,
+      const size_t poolSize,
+      const double slabPenalty, 
+      double radius):
+      visibleSize(visibleSize),
+      hiddenSize(hiddenSize),
+      poolSize(poolSize),
+      slabPenalty(slabPenalty),
+      radius(radius)
+{
+  parameter.set_size(visibleSize * hiddenSize * poolSize +
+      visibleSize + hiddenSize);
+
+  assert(slabPenalty.n_rows == 1);
+  assert(slabPenalty.n_cols == 1);
+
+  visibleMean.set_size(visibleSize, 1);
+  spikeMean.set_size(hiddenSize, 1);
+  spikeSamples.set_size(hiddenSize, 1);
+  slabMean.set_size(poolSize, hiddenSize);
+};
+
+// Reset function
+inline void SpikeSlabRBMPolicy::Reset()
+{
+  // Weight shape D * K * N
+  weight = arma::cube(parameter.memptr(), visibleSize, poolSize, hiddenSize,
+      false, false);
+
+  // spike bias shape N * 1
+  spikeBias = arma::mat(parameter.memptr() + weight.n_elem, hiddenSize, 1,
+      false, false);
+
+  // visible penalty 1 * 1 => D * D(when used)
+  visiblePenalty = arma::mat(parameter.memptr() + weight.n_elem +
+      spikeBias.n_elem, 1, 1, false, false);
+
+  diagInvSlabPenalty = 1.0 / slabPenalty * arma::eye(poolSize, poolSize);
+  diagSlabPenalty = slabPenalty * arma::eye(poolSize, poolSize);
+}
+
+/**
+ * Free energy of the spike and slab variable
+ * the free energy of the ssRBM is given my
+ * $v^t$$\Delta$v - $\sum_{i=1}^N$ 
+ * $\log{ \sqrt{\frac{(-2\pi)^K}{\prod_{m=1}^{K}(\alpha_i)_m}}}$ -
+ * $\sum_{i=1}^N \log(1+\exp( b_i +
+ * \sum_{m=1}^k \frac{(v(w_i)_m^t)^2}{2(\alpha_i)_m})$
+ *
+ * @param input the visible layer
+ */ 
+inline double SpikeSlabRBMPolicy::FreeEnergy(arma::mat&& input)
+{
+  assert(input.n_rows == visibleSize);
+  assert(input.n_cols == 1);
+
+  scalarVisiblePenalty = visiblePenalty(0, 0);
+
+  double freeEnergy = 0.5 * arma::as_scalar(input.t() *
+      scalarVisiblePenalty * input);
+
+  assert(std::isfinite(std::log(2.0 * M_PI / slabPenalty)));
```

Comment:
I think this is not needed since `slabPenalty` is a constant. Right now (after I fixed `SampleVisible()` and `MeanVisible()`) I can't reproduce errors with NANs and INFs. So, it seems the error has already been fixed.